# Manejo estrategico de los datos de Global-Tech Retai
Este proyecto busca aprovechar los datos recolectados por Global-Tech Retail como base para un análisis profundo, con el objetivo de generar recomendaciones estratégicas para la empresa.

Creado por: Edgar Yesid Rentería Rojas  
Julio 2024

# Importamos librerias

In [323]:
#*! RECUERDEN instalarlos poninendo pip install nombre_del_paquete en la terminal de vscode
import pandas as pd
import numpy as np
import pandas as pd
import sqlite3
from sklearn.preprocessing import MinMaxScaler, StandardScaler

#import seaborn as sn  #*!este no lo vamos a usar por ahora
#import matplotlib.pyplot as plt

## 🖇️Conexion con base de datos y creacion de DF

In [324]:
# Conectarse al archivo .db
conexion = sqlite3.connect('C:\\Users\\edgar\\OneDrive - Universidad Nacional de Colombia\\Analisis de factos\\My-local-learning-path\\My-learning-path\\Proyecto Python\\Documentos\\clientes.db')

# Consultar las tabla
consultar = 'SELECT * from sqlite_master'

print("🗂️ Consulta de sqlite_master:", pd.read_sql_query(consultar, conexion),'\f \f')

# Leer y asignar una tabla específica
clientes_df = pd.read_sql_query("SELECT * FROM clientes;", conexion)

#? ahora ponemos usar comandos de pandas para manipular data como:
#? 
conexion.close()

#! LEEMOS LOS CSV QUE NECESITAMOS PARA EL PROYECTO
productos_df = pd.read_csv('C:\\Users\\edgar\\OneDrive - Universidad Nacional de Colombia\\Analisis de factos\\My-local-learning-path\\My-learning-path\\Proyecto Python\\Documentos\\productos.csv')

ventas_df = pd.read_csv('C:\\Users\\edgar\\OneDrive - Universidad Nacional de Colombia\\Analisis de factos\\My-local-learning-path\\My-learning-path\\Proyecto Python\\Documentos\\ventas.csv')


🗂️ Consulta de sqlite_master:     type      name  tbl_name  rootpage  \
0  table  clientes  clientes         2   

                                                 sql  
0  CREATE TABLE "clientes" (\n"ID_Cliente" INTEGE...    


--- 
# visualizacion basica

## 📦Datos de productos

In [325]:
print("\n📦 DATOS DE PRODUCTOS\n")

# 🧭 Muestra la cabecera y los primeros 10 datos del DataFrame
print("🔹 Cabecera y primeros 10 registros:\n")
print(productos_df.value_counts().head(11), "\n")

# 📊 Muestra un resumen estadístico del DataFrame
print("🔹 Resumen estadístico de productos:\n")
print(productos_df.describe(include='all'), "\n")

# ℹ️ Muestra información general del DataFrame
print("🔹 Información del DataFrame productos:\n")
productos_df.info()
print("\n")

'''top_10_productos = productos_df['Nombre_Producto'].value_counts().head(10)
print(top_10_productos)'''

'''top_10_productos_costo = productos_df['Costo_Unitario'].value_counts().head(10)
print(top_10_productos_costo)'''


📦 DATOS DE PRODUCTOS

🔹 Cabecera y primeros 10 registros:

ID_Producto  Nombre_Producto               Categoria    Costo_Unitario
109          Surface Laptop 4              Laptops      894.82            2
124          Motorola Edge 40              Smartphones  358.03            2
120          Xiaomi Mi 13                  Smartphones  675.25            2
151          Mouse Pad XXL                 Accesorios   85.58             2
162            Altavoces Logitech          Audio        156.92            2
158          Parlante JBL Charge 5         Audio        188.31            2
156          Auriculares Sony WH-1000XM5®  Audio        296.81            2
104          Dell XPS 13                   Laptops      559.07            1
105            Asus ZenBook 14             Laptops      1047.54           1
107          HP Pavilion                   Laptops      477.37            1
106          Lenovo IdeaPad                Laptops      672.20            1
Name: count, dtype: int64 

🔹 Res

"top_10_productos_costo = productos_df['Costo_Unitario'].value_counts().head(10)\nprint(top_10_productos_costo)"

### 📦 Nuestros productos informe preliminar Sobre 💻🛒

Estimados miembros del equipo,  
les presentamos un informe sobre los datos de nuestros productos.
Esperamos que lo encuentren tan bueno como una arepa de chocolo recién salida del fogón. 🌽🔥

---

#### 🌟 Principales Conclusiones

- 🔢 Se identifican **71 productos distintos** en el catálogo.  
- 💻 El líder en ventas es el **MacBook Pro 14"**, destacándose en todos los segmentos.  
- 🏷️ La categoría más representativa es **Laptop**, con un desempeño constante.

---

#### 📈 Resumen de Precios

- ☕ Precio promedio: **\$376.03** (para referencias comparativas).  
- 🎯 Mediana (50 %): **\$296.34**, indicador del punto medio de la distribución.  
- 🔢 Rango intercuartílico (25 %–75 %): **\$192.52 – \$546.60**, reflejo de la variabilidad central.  
- 🏔️ Mínimo: **\$24.79**, valor de entrada al segmento.  
- 🏔️ Máximo: **\$1 047.54**, tope de la oferta premium.

---

#### ⚠️ Calidad de Datos

- 🕵️‍♂️ Existen **5 registros** sin precio asignado (7 % del total); se sugiere completar o validar estas entradas.  
- 🔄 Se detectan **duplicados**: 63 nombres únicos en 71 registros.  
  - Por ejemplo, **MacBook Pro 14"** aparece en dos filas; conviene verificar su origen.

---

#### 📊 Dispersión de Precios

- 🌄 Desviación estándar: **\$259.07**  
  - Esto indica que, en promedio, los precios se alejan \$259.07 del valor medio.  
  - El rango evidencia diversidad, desde opciones económicas hasta equipos de alta gama.

---

Esperamos que este análisis les sea de gran utilidad para orientar la estrategia de catálogo y márgenes de rentabilidad.  

Atentamente,  
El equipo de Análisis Estratégico 😎🤝🏽  

## 🧔🏽Base de Clientes  : Informe Preliminar

In [326]:
print("\n🗂️ DATOS DE CLIENTES\n")

# 🧭 Muestra la cabecera y los primeros 10 datos del DataFrame
print("🔹 Cabecera y primeros 10 Datos de clientes:\n")
print(clientes_df.value_counts().head(11), "\n")

# 📊 Muestra un resumen estadístico del DataFrame
print("🔹 Resumen estadístico de clientes:\n")
print(clientes_df.describe(include='all'), "\n")

# ℹ️ Muestra información general del DataFrame
print("🔹 Información del DataFrame clientes:\n")
clientes_df.info()
print("\n")
print(f" Primeros 20 Datos de clientes:", clientes_df.head(20),'\n \n')


🗂️ DATOS DE CLIENTES

🔹 Cabecera y primeros 10 Datos de clientes:

ID_Cliente  Ciudad       Pais       Fuente_Marketing 
825         Manta        Ecuador    Redes Sociales       1
1           Chiclayo     Peru       Telemarketing        1
2             Rosario    Argentina  Email Marketing      1
3           La Plata     Argentina  Email Marketing      1
4           Rancagua     Chile      Referido             1
5             Durán      Ecuador    YouTube              1
6           Medellín     Colombia   WhatsApp Business    1
809         Sevilla      España     YouTube              1
808         Santiago     Chile      Redes Sociales       1
807         Lima         Peru       Influencers          1
806         Santa Fe     Argentina  Eventos              1
Name: count, dtype: int64 

🔹 Resumen estadístico de clientes:

        ID_Cliente    Ciudad    Pais Fuente_Marketing
count    825.00000       825     811              805
unique         NaN       128      16               23
top

Emprendemos un recorrido estratégico por los datos de nuestra comunidad de clientes, revelando patrones y oportunidades con la precisión de un café antioqueño bien colado en la mañana.

---

### 🗺️ Panorama General

- Total de clientes registrados: **825**  
- Ciudades representadas: **128**  
- Países cubiertos: **16**  
- Canales de Marketing activos: **23**

Nuestra comunidad es diversa, con picos de actividad en México y Argentina, y un mosaico de ciudades que nos regalan perspectivas locales.

---

### 🔍 Insights Clave

- La banda intercuartílica de IDs (207–619) muestra un flujo constante de adquisiciones.  
- Valencia, Rosario y Santa Fe encabezan la lista de urbes con más conexiones.

Cada dato nos revela un rincón distinto de nuestro territorio comercial.

---

### ⚠️ Calidad de Datos

- Falta información de “País” en **14** registros (~1.7 %).  
- “Fuente_Marketing” ausente en **20** casos (~2.4 %).  
- No hay duplicados de ID_Cliente: la llave primaria está impecable.

Antes de segmentar, conviene homogeneizar y completar estos campos para evitar senderos mal trazados.

---

### 📍 Huellas Geográficas

- México y Argentina concentran cerca del **30 %** de nuestra clientela.  
- Valencia lidera con **22** conexiones, seguida de Rosario y Santa Fe.  
- El resto de ciudades aporta un mosaico de micromercados con gran potencial de exploración.

Nuestros pasos se extienden por todo el continente, desde Chiclayo hasta Medellín.

---

### 📣 Voces del Marketing

- Influencers atrae al **9.1 %** de la audiencia (75 clientes).  
- Email Marketing y Telemarketing mantienen un pulso constante.  
- YouTube y Referidos registran picos interesantes en segmentos puntuales.

La cúspide del engagement pasa por las recomendaciones y la credibilidad de terceros.

---

### 🎯 Recomendaciones de Altura

1. Estandarizar y rellenar campos faltantes para afinar segmentaciones.  
2. Priorizar campañas en México y Argentina, donde nace casi un tercio de la demanda.  
3. Concentrar recursos en las 5 ciudades líderes para maximizar retorno.  
4. Fortalecer alianzas con Influencers y potenciar micro-influencers regionales.  
5. Ejecutar auditoría de costos por canal para optimizar el presupuesto de captación.

---

El equipo de Análisis Estratégico te acompaña en cada ascenso. 😎🤝🏽  

## 🛍️ DATOS DE VENTAS
### Analisis preliminar de tabla ventas 


In [327]:
print("\n🛍️ DATOS DE VENTAS\n")

# 🧭 Muestra la cabecera y los primeros 10 datos del DataFrame
print("🔹 Cabecera y primeros 10 Datos de ventas:\n")
print(ventas_df.value_counts().head(11), "\n")

# 📊 Muestra un resumen estadístico del DataFrame
print("🔹 Resumen estadístico de ventas:\n")
print(ventas_df.describe(include='all'), "\n")

# ℹ️ Muestra información general del DataFrame
print("🔹 Información del DataFrame ventas:\n")
ventas_df.info()
#print("\n")
#print(f" Primeros 20 Datos de ventas:", ventas_df.head(20),'\n \n')


🛍️ DATOS DE VENTAS

🔹 Cabecera y primeros 10 Datos de ventas:



ID_Venta  ID_Producto  ID_Cliente  Fecha_Venta  Cantidad  Precio_Unitario
12230     122          140         2022-09-09   4         811.15             1
1         157          175         2021-11-27   5         305.31             1
2         119          708         2021-01-23   1         600.98             1
3         155          663         2021-04-08   1         117.99             1
4         153          1014        2023-01-03   2         42.19              1
5         146          711         2023-04-09   2         52.07              1
6         139          333         2022-11-01   4         310.76             1
7         113          334         2023-07-02   2         735.49             1
8         106          45          2023-01-26   4         921.83             1
9         108          259         2021-12-18   2         1077.65            1
12214     112          404         2023-06-25   2         859.11             1
Name: count, dtype: int64 

🔹 Resumen estadístico de vent

A continuación, se presentan las conclusiones preliminares tras un primer análisis del dataset de ventas.

---

## 1. Calidad de datos

- Fechas con formato inválido (`2025-13-45`) indican registros erróneos.  
- Existen valores negativos en `Cantidad` (mínimo –1) y `Precio_Unitario` (mínimo –99.91).  
- 268 registros (~2.2%) tienen `Precio_Unitario` nulo.  

---

## 2. Distribución de Cantidad

- **Media**: 11.95 unidades por venta  
- **Mediana**: 3 unidades  
- **Rango intercuartil (Q1–Q3)**: 2–4 unidades → la mayoría de ventas son pedidos pequeños  
- **Outliers**: pedidos muy grandes (hasta 997 unidades) o negativos → conviene revisar devoluciones y errores de captura  

---

## 3. Distribución de Precio Unitario

- **Media**: \$702.25  
- **Mediana**: \$417.92  
- **Rango intercuartil (Q1–Q3)**: \$258.84–\$722.64 → la mayoría de productos se sitúa en precios medios  
- **Valores extremos**: hasta \$49,256 y negativos → pueden corresponder a descuentos, ajustes manuales o errores  

---

## 4. Clientes y productos

- 75% de los productos tienen `ID_Producto` ≤ 148  
- 75% de los clientes tienen `ID_Cliente` ≤ 626  
- Presencia de IDs muy altos sugiere registros atípicos o incorporaciones recientes  

---

## 5. Próximos pasos sugeridos

1. **Limpiar datos**  
   - Corregir o descartar fechas inválidas  
   - Filtrar/ajustar cantidades y precios negativos  
   - Imputar o eliminar precios nulos  

2. **Calcular ingresos** (`Cantidad × Precio_Unitario`) y analizar su distribución  

3. **Visualizar la evolución mensual y anual de ingresos**  

4. **Identificar los Top 10 productos y clientes por facturación**  

---

# 😈Funcion normalisadora🔥 
Creamos una funcion bien poderosa que nos permita normalizar df de tacazo 
Esto se hace para facilitar la rteutilizacion del codigo

In [328]:
# 😈Funcion normalisadora🔥

def normalizar_df(
    df: pd.DataFrame,
    drop_duplicates: bool = True,
    impute_strategy: str = None,        # 'mean', 'median', 'mode' o None (dropna)
    cols_impute: list = None,           # columnas para imputar
    cols_positive: list = None,         # columnas que deben ser > 0
    scaler: str = 'minmax',             # 'minmax' o 'standard'
    cols_scale: list = None,             # columnas a escalar
    drop_columns: list = None,         # columnas a eliminar
    change_types: dict = None,             # diccionario con columnas y tipos deseados
) -> pd.DataFrame:
    """
    🧼 Limpia y escala un DataFrame de forma configurable y melísima.

    Parámetros
    ----------
    df : pd.DataFrame
        DataFrame original que necesita una manito.
    drop_duplicates : bool
        Si es True, elimina filas duplicadas.
    impute_strategy : str o None
        Estrategia para rellenar nulos:
        -Media: útil para datos numéricos simétricos.

        -Mediana: mejor para datos con outliers.

        -Moda: ideal para datos categóricos.
        - None     : elimina filas con cualquier nulo
    cols_impute : list de str
        Columnas donde aplicar la imputación.
    cols_positive : list de str
        Columnas que deben tener valores estrictamente positivos (> 0).
    scaler : str
        Tipo de escalado:
        - 'minmax'   : lleva los valores al rango [0, 1]
        - 'standard' : centra en cero y escala con desviación estándar
    cols_scale : list de str
        Columnas numéricas que se deben escalar.

    Retorna
    -------
    pd.DataFrame
        El DataFrame limpio, filtrado y escalado, listo pa’ modelar.
    """
    # 🧼 Limpieza de nombres de columnas
    df.columns = df.columns.str.strip().str.replace(' ', '_')

    # 🧹 Paso extra: eliminar columnas innecesarias
    if drop_columns:
        df.drop(columns=[col for col in drop_columns if col in df.columns], inplace=True)

    # 🧪 Trabajamos sobre una copia para no dañar el original
    df = df.copy()

    # 🔄 Cambio de tipo de datos
    if change_types:
        for col, dtype in change_types.items():
            if col in df.columns:
                try:
                    df[col] = df[col].astype(dtype)
                except Exception as e:
                    print(f"⚠️ No se pudo convertir '{col}' a {dtype}: {e}")

    # Validar columnas para imputar si corresponde
    if cols_impute is not None:
        for col in cols_impute:
            if col not in df.columns:
                raise KeyError(f"La columna '{col}' no existe en el DataFrame.")

    # 🔁 Paso 1: Eliminar duplicados
    if drop_duplicates:
        df.drop_duplicates(inplace=True)
    # 🧼 Paso 2: Manejo de nulos
    if impute_strategy and cols_impute:
        for col in cols_impute:
            if impute_strategy == 'mean':
                df[col] = df[col].fillna(df[col].mean())
            elif impute_strategy == 'median':
                df[col] = df[col].fillna(df[col].median())
            elif impute_strategy == 'mode':
                df[col] = df[col].fillna(df[col].mode()[0])
            else:
                raise ValueError("Estrategia de imputación no válida. Usa 'mean', 'median' o 'mode'.")
    else:
        df.dropna(inplace=True)
    # 🚫 Paso 3: Filtrar valores no positivos
    if cols_positive:
        for col in cols_positive:
            df = df[df[col] > 0]

    # 📊 Paso 4: Escalado de columnas numéricas
    if cols_scale:
        if scaler == 'minmax':
            _scaler = MinMaxScaler()
        elif scaler == 'standard':
            _scaler = StandardScaler()
        else:
            raise ValueError("Escalador no soportado. Usa 'minmax' o 'standard'.")

        df[cols_scale] = _scaler.fit_transform(df[cols_scale])

    return df


# 🧼 Normalización de Tablas: Ventas, Productos y Clientes


Este documento explica las decisiones tomadas para limpiar y normalizar las tablas del proyecto, con el objetivo de garantizar coherencia lógica, facilitar el análisis y preparar los datos para modelos o visualizaciones.

---

## 🛍️ Ventas

**🔧 Decisiones tomadas:**
- Se imputan valores nulos en `Precio_Unitario` usando la **mediana**, ya que es menos sensible a outliers y mantiene estabilidad en precios variables.
- Se eliminan valores negativos en `Cantidad` y `Precio_Unitario`, porque no tiene sentido registrar ventas con cantidades o precios negativos.
- Se aplica **MinMaxScaler** para escalar estas variables entre 0 y 1, lo que facilita comparaciones y análisis visuales.

**🎯 Objetivo:**  
Preparar los datos para análisis de ingresos, comportamiento de ventas y evitar distorsiones por valores extremos.

In [ ]:
ventas_df_limpio['Ingreso_Total'] = ventas_df_limpio['Cantidad'] * ventas_df_limpio['Precio_Unitario']
print(ventas_df_limpio[['Cantidad', 'Precio_Unitario', 'Ingreso_Total']].head(10))
# columna a añadir

    Cantidad  Precio_Unitario  Ingreso_Total
0   0.004016         0.013282       0.000053
2   0.000000         0.001376       0.000000
3   0.001004         0.004742       0.000005
4   0.002008         0.012729       0.000026
5   0.003012         0.021718       0.000065
7   0.004016         0.014091       0.000057
8   0.001004         0.022093       0.000022
9   0.003012         0.012602       0.000038
10  0.000000         0.011931       0.000000
12  0.004016         0.015754       0.000063


In [ ]:
# Normalización de la tabla de ventas
ventas_df['Fecha_Venta'] = pd.to_datetime(ventas_df['Fecha_Venta'], errors='coerce')

ventas_df_limpio = normalizar_df(
    df=ventas_df,
    drop_duplicates=True,
    impute_strategy='median',
    cols_impute=['Precio_Unitario'],
    cols_positive=['Cantidad', 'Precio_Unitario'],
    scaler='minmax',
    cols_scale=['Cantidad', 'Precio_Unitario'],
    change_types={'Fecha_Venta': 'datetime64[ns]'},  # Convertimos la fecha a tipo datetime
    drop_columns=['ID_Venta']  # Eliminar columna 'id' si no es necesaria
)
print("🛍️ Ventas normalizadas:\n", ventas_df_limpio.head(10), '\n \n')
print("🛍️ Ventas normalizadas info:\n", ventas_df_limpio.info(), '\n \n')

🛍️ Ventas normalizadas:
     ID_Producto  ID_Cliente Fecha_Venta  Cantidad  Precio_Unitario
0           110          57  2023-01-24  0.004016         0.013282
2           152         743  2023-10-29  0.000000         0.001376
3           158         283  2022-09-21  0.001004         0.004742
4           104         396  2021-08-18  0.002008         0.012729
5           108         579  2021-08-01  0.003012         0.021718
7           111         367  2023-08-24  0.004016         0.014091
8           101         684  2022-08-10  0.001004         0.022093
9           142           3  2021-10-14  0.003012         0.012602
10          141         582  2023-11-10  0.000000         0.011931
12          113         687  2023-03-22  0.004016         0.015754 
 

<class 'pandas.core.frame.DataFrame'>
Index: 11158 entries, 0 to 12229
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   ID_Producto      111

---

## 📦 Productos

**🔧 Decisiones tomadas:**
- Se imputan valores nulos en `Costo_Unitario` y `Precio_Venta` con la **mediana**, para mantener robustez frente a valores atípicos.
- Se eliminan valores negativos en ambas columnas, ya que un producto no puede tener costo o precio negativo.
- Se usa **StandardScaler** para centrar y escalar los datos, útil para análisis financiero, clustering o reducción de dimensionalidad (PCA).

**🎯 Objetivo:**  
Garantizar consistencia financiera y preparar los datos para análisis técnico o segmentación de productos.

In [330]:
# Normalización de la tabla de productos
productos_df_limpio = normalizar_df(
    df=productos_df,
    drop_duplicates=True,
    impute_strategy='mode',
    cols_impute=['Categoria', 'Nombre_Producto'],
    cols_positive=['Costo_Unitario'],
    scaler='standard',
    cols_scale=['Costo_Unitario', 'Costo_Unitario'],
    drop_columns=['ID_Producto'],  # Eliminar columna 'id' si no es necesaria
)
print("👤 Clientes normalizados:\n", productos_df_limpio.head(10), '\n \n')


👤 Clientes normalizados:
           Nombre_Producto Categoria  Costo_Unitario
0     Laptop ProBook G9     Laptops        2.103442
1     Laptop ThinkPad T14   Laptops        2.408181
3             Dell XPS 13   Laptops        0.719989
4       Asus ZenBook 14     Laptops        2.637246
5          Lenovo IdeaPad   Laptops        1.164027
6             HP Pavilion   Laptops        0.399314
7           Acer Aspire 5   Laptops        1.779392
8        Surface Laptop 4   Laptops        2.037816
9          MSI Creator 15   Laptops        0.811245
10         Razer Blade 15   Laptops        0.968482 
 



---

## 👥 Clientes

**🔧 Decisiones tomadas:**
- Se eliminan **duplicados** para evitar registros repetidos que puedan distorsionar métricas de comportamiento o segmentación.
- No se imputan ni escalan columnas, ya que no hay variables numéricas relevantes que lo justifiquen.

**🎯 Objetivo:**  
Mantener una base limpia y única de clientes sin aplicar transformaciones innecesarias.

In [ ]:

# Normalización de la tabla de clientes
clientes_df_limpio = normalizar_df(
    df=clientes_df,
    drop_duplicates=True,
    impute_strategy=None,
    cols_impute=None,
    cols_positive=None,
    scaler='minmax',
    cols_scale=None,
    drop_columns=['ID_Cliente'],  # Eliminar columna 'id' si no es necesaria
)
print("👤 Clientes normalizados:\n", clientes_df_limpio.head(10), '\n \n')

👤 Clientes normalizados:
         Ciudad       Pais   Fuente_Marketing
0     Chiclayo       Peru      Telemarketing
1    Rosario    Argentina    Email Marketing
2     La Plata  Argentina    Email Marketing
3     Rancagua      Chile           Referido
4      Durán      Ecuador            YouTube
5     Medellín   Colombia  WhatsApp Business
6     Santiago      Chile            Eventos
7        Quito    Ecuador      Telemarketing
8      Iquitos       Peru            Eventos
9    La Serena      Chile            YouTube 
 

👤 Clientes normalizados:
 <bound method NDFrame.describe of                    Ciudad       Pais Fuente_Marketing
0                Chiclayo       Peru    Telemarketing
1               Rosario    Argentina  Email Marketing
2                La Plata  Argentina  Email Marketing
3                Rancagua      Chile         Referido
4                 Durán      Ecuador          YouTube
..                    ...        ...              ...
790    Ciudad de México       Mexico 